File: java/src/test/java/com/penapereira/patterns/builder/BuilderTest.java
package com.penapereira.patterns.builder;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

class BuilderTest {

    @Test
    void theDirectorFixesTheSequence() {
        assertEquals("Product(PartA, PartB)", new Director().construct(new ConcreteBuilder()).describe());
    }

    @Test
    void aBuilderAloneProducesOnlyTheRequestedParts() {
        ConcreteBuilder builder = new ConcreteBuilder();
        assertEquals("Product()", builder.getResult().describe());
        builder.buildPartB();
        assertEquals("Product(PartB)", builder.getResult().describe());
    }

    @Test
    void theSameDirectorDrivesADifferentRepresentation() {
        Builder upperCase = new Builder() {
            private final Product product = new Product();

            @Override
            public void buildPartA() {
                product.add("A");
            }

            @Override
            public void buildPartB() {
                product.add("B");
            }

            @Override
            public Product getResult() {
                return product;
            }
        };
        assertEquals("Product(A, B)", new Director().construct(upperCase).describe());
    }
}

File: javascript/src/builder/builder.test.js
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { BufferOutput } from "../runtime/contract.js";
import { ConcreteBuilder, Director, Product } from "./builder.js";
import { builderExample } from "./example.js";

describe("Builder", () => {
  it("the director fixes the sequence", () => {
    assert.equal(new Director().construct(new ConcreteBuilder()).describe(), "Product(PartA, PartB)");
  });

  it("a builder alone produces only the requested parts", () => {
    const builder = new ConcreteBuilder();
    assert.equal(builder.getResult().describe(), "Product()");
    builder.buildPartB();
    assert.equal(builder.getResult().describe(), "Product(PartB)");
  });

  it("the same director drives a different representation", () => {
    const product = new Product();
    const shortNames = {
      buildPartA: () => product.add("A"),
      buildPartB: () => product.add("B"),
      getResult: () => product,
    };
    assert.equal(new Director().construct(shortNames).describe(), "Product(A, B)");
  });

  it("example prints the expected lines", () => {
    const out = new BufferOutput();
    builderExample.run(out);
    assert.equal(builderExample.id, "builder");
    assert.deepEqual(out.lines, [
      "Executing Builder Pattern Implementation",
      "  Director.construct(ConcreteBuilder): Product(PartA, PartB)",
      "  ConcreteBuilder alone, only part B: Product(PartB)",
    ]);
  });
});
